### PR TITLE
Support for np.trapz

### PR DESCRIPTION
**Pull request by rjenc29**
Per #3442

Support for first 3 arguments of `np.trapz`

Work in progress - initial commit for CI.  Items left to do include:

- [x] Update docs
- [x] Simplify implementation
- [x] Negative tests
- [x] Sensible exception messages

### CODE REVIEW

**Comment by rjenc29** [comment]
Hi, when you have a mo, please let me know how you feel about this PR.

There's quite a lot of undocumented behaviour (e.g. where arguments are multi-dimensional; where inputs are complex etc) which I attempted to figure out through testing.  

It's ended up quite branch-heavy & have done my best to ensure a given input hits the most appropriate branch for the majority of cases (although I am sure there are a few gaps and some of the function names I've used might be a bit misleading).

There's one test case which fails, but I think NumPy behaviour is questionable, so I have captured it and marked it to skip; let me know what you reckon.  

Aside the unit tests, I've used it for a few 'real world' numerical integrations and it seems like a good speedup over NumPy.

Cheers!

**Comment by stuartarchibald** [comment]
Thanks very much for doing the work on this. I've taken a look and see what you mean by the branch heaviness, inspired by https://github.com/numpy/numpy/blob/7ee52003a8be7d4d83c8837d7a795f759b0a7ba2/numpy/lib/function_base.py#L4040-L4065 I've written https://github.com/stuartarchibald/numba/commit/4f453b50d11374312a392e5b940ccc6a31514077 which you are welcome to use if useful, though it doesn't contain the short cutting for easily identified cases or exception handling yet, and I'm not sure of the performance.

**Comment by rjenc29** [comment]
Thanks very much for the feedback - I will have a proper look over the coming weekend and revert.

**Comment by rjenc29** [comment]
Yeah, that works a treat - nice one.  I'm not sure I would have come up with that one, even with infinite caffeine and sugar.

Performance wise, it seems fine.  The only branch which was consistently a bit quicker in the old branch-heavy implementation was where it ended up doing `np.sum(y_ave) * dx_scalar`.  

(The branch which hit `np.dot(y_ave, dx_array)` was consistently the slowest.)

So, on balance, all the tests pass; it performs well and the implementation's a small amount of quite readable code (albeit not code I can really claim authorship of) :)

I put in a couple of error handling checks and a few more tests.

Please let me know what you think.

File: numba/tests/test_np_functions.py [hunk 1]
```diff
@@ -1590,6 +1602,231 @@ def test_roll_exceptions(self):
...
+    def test_trapz_numpy_questionable(self):
+        pyfunc = np_trapz
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc)
+
+        # passes (NumPy and Numba return 2.0)
+        y = np.array([True, False, True, True]).astype(np.int)
+        _check({'y': y})
+
+        # fails (NumPy returns 1.5; Numba returns 2.0)
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: Perhaps report this upstream to NumPy/see if it's on their issue tracker? Having just drawn a picture, I'm pretty sure the answer should be 2.0.

**Comment by seibert** [comment]
This looks good to me.  Thanks for the work on refining the implementation!

**Comment by stuartarchibald** [comment]
Thanks @seibert

**Pull request merged**